### PR TITLE
Bug #40 fix: Bug fixes against the SQL injection in form fields

### PR DESCRIPTION
**Pull request by sagar-ttpl**
Bug #40 fix: Bug fixes against the SQL injection in form fields

### CODE REVIEW

**Comment by sagar-ttpl** [comment]
This PR contains the bug fixes against the malicious inputs and contains the fixes done against the code open to XSS attacks.

File: site/helpers/tjfields.php [hunk 1]
```diff
@@ -693,50 +709,41 @@ public static function buildFilterModuleQuery()
...
-				 * SELECT fv1. *
-					FROM  `xcqpa_tjfields_fields_value` AS fv1
-					INNER JOIN  `xcqpa_tjfields_fields_value` AS fv2 ON fv2.`content_id` = fv1.content_id
-					WHERE fv1.option_id
-					IN ( 18 ) AND
-					WHERE fv2.option_id
-					IN ( 13,14 )
-				 * */
-				$query->select('DISTINCT fv1.content_id');
+				$query->select($db->quoteName('DISTINCT fv1.content_id'));
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: DISTINCT keyword inside quoteName will cause issues, re-test this once.

File: site/helpers/tjfields.php [hunk 2]
```diff
@@ -693,50 +709,41 @@ public static function buildFilterModuleQuery()
 				{
 					if (!empty($fFieldAndFieldOptions))
 					{
-						$query->select('DISTINCT fv1.content_id');
-						$query->from('#__tjfields_fields_value AS fv1');
-						$query->where("fv1.option_id IN (" . $fFieldAndFieldOptions->optionsStr . ")");
+						$query->select($db->quoteName('DISTINCT fv1.content_id'));
+						$query->from($db->quoteName('#__tjfields_fields_value', 'fv1'));
+						$query->where($db->quoteName('fv1.option_id' . 'IN (' . $db->quote($fFieldAndFieldOptions->optionsStr) . ')'));
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: $fFieldAndFieldOptions => $fieldAndFieldOptions OR maybe some more relivent name

File: site/helpers/tjfields.php [hunk 3]
```diff
@@ -693,50 +709,41 @@ public static function buildFilterModuleQuery()
 				{
 					if (!empty($fFieldAndFieldOptions))
 					{
-						$query->select('DISTINCT fv1.content_id');
-						$query->from('#__tjfields_fields_value AS fv1');
-						$query->where("fv1.option_id IN (" . $fFieldAndFieldOptions->optionsStr . ")");
+						$query->select($db->quoteName('DISTINCT fv1.content_id'));
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: DISTINCT in quoteName will cause unexpected results, re-test this and update in all places

File: site/helpers/tjfields.php [hunk 4]
```diff
@@ -554,30 +554,37 @@ public function getFilterableFields($client, $category_id = '')
...
+				$query->join('INNER', '#__tjfields_category_mapping AS fcm ON fcm.field_id = f.id');
 				$query->where('fcm.category_id=' . $category_id);
 			}
-			else*/
+			else
 
-			{
+			{*/
 				// Doesn't have mapped any category
 				$query->where('NOT EXISTS (select * FROM #__tjfields_category_mapping AS cm where f.id=cm.field_id)');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: query builder for sub query

File: site/helpers/tjfields.php [hunk 5]
```diff
@@ -554,30 +554,37 @@ public function getFilterableFields($client, $category_id = '')
...
+				'NOT EXISTS (select * FROM' . $db->quoteName('#__tjfields_category_mapping', 'cm')
+				. ' where ' . $db->quoteName('f.id')
+				. ' = '
+				. $db->quoteName('cm.field_id')
+			);
 
 			/*if (!empty($category_id))
 			{
-				$query->JOIN('INNER', '#__tjfields_category_mapping AS fcm ON fcm.field_id = f.id');
+				$query->join('INNER', '#__tjfields_category_mapping AS fcm ON fcm.field_id = f.id');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use query builder to construct query

File: site/helpers/tjfields.php [hunk 6]
```diff
@@ -554,30 +554,37 @@ public function getFilterableFields($client, $category_id = '')
...
+			$query->join('INNER', $db->quoteName('#__tjfields_options', 'fo') . 'ON' . $db->quoteName('fo.id') . ' = ' . $db->quoteName('fv.option_id'));
+			$query->where($db->quoteName('f.client') . ' = ' . $db->quote($client));
+			$query->where($db->quoteName('f.filterable') . ' = 1');
+			$query->where($db->quoteName('f.state') . '=1');
+			$query->where($db->quoteName('fv.option_id') . ' IS NOT NULL');
+			$query->where($db->quoteName('f.type') . 'IN ("single_select", "multi_select", "radio")');
+
+			// Doesn't have mapped any category
+			$query->where(
+				'NOT EXISTS (select * FROM' . $db->quoteName('#__tjfields_category_mapping', 'cm')
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use query builder to construct sub query

File: site/helpers/tjfields.php [hunk 7]
```diff
@@ -554,30 +554,37 @@ public function getFilterableFields($client, $category_id = '')
...
-			$query->where("f.type IN ('single_select','multi_select', 'radio')");
+			$query->select($db->quoteName(array('DISTINCT fv.option_id', 'f.id', 'f.name', 'f.label', 'fv.value', 'fo.options')));
+			$query->from($db->quoteName('#__tjfields_fields', 'f'));
+			$query->join('INNER', $db->quoteName('#__tjfields_fields_value', 'fv') . 'ON' . $db->quoteName('fv.field_id') . ' = ' . $db->quoteName('f.id'));
+			$query->join('INNER', $db->quoteName('#__tjfields_options', 'fo') . 'ON' . $db->quoteName('fo.id') . ' = ' . $db->quoteName('fv.option_id'));
+			$query->where($db->quoteName('f.client') . ' = ' . $db->quote($client));
+			$query->where($db->quoteName('f.filterable') . ' = 1');
+			$query->where($db->quoteName('f.state') . '=1');
+			$query->where($db->quoteName('fv.option_id') . ' IS NOT NULL');
+			$query->where($db->quoteName('f.type') . 'IN ("single_select", "multi_select", "radio")');
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: quote for strings inside IN clause

File: site/helpers/tjfields.php [hunk 8]
```diff
@@ -554,30 +554,37 @@ public function getFilterableFields($client, $category_id = '')
...
-			$query->FROM("#__tjfields_fields AS f");
-			$query->JOIN('INNER', '#__tjfields_fields_value AS fv ON fv.field_id = f.id');
-			$query->JOIN('INNER', '#__tjfields_options AS fo ON fo.id = fv.option_id');
-
-			$query->where('f.client="' . $client . '"');
-			$query->where('f.filterable=1');
-			$query->where('f.state=1');
-			$query->where('fv.option_id IS NOT NULL');
-			$query->where("f.type IN ('single_select','multi_select', 'radio')");
+			$query->select($db->quoteName(array('DISTINCT fv.option_id', 'f.id', 'f.name', 'f.label', 'fv.value', 'fo.options')));
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: DISTINCT inside quoteName will result into unexpected results

File: site/helpers/tjfields.php [hunk 9]
```diff
@@ -505,10 +496,18 @@ public function getUniversalFields($client)
...
 			$db    = JFactory::getDbo();
 			$query = $db->getQuery(true);
-			$query->select('DISTINCT * FROM #__tjfields_fields AS f');
-			$query->where('NOT EXISTS (select * FROM #__tjfields_category_mapping AS cm where f.id=cm.field_id)');
-			$query->where('f.client="' . $client . '"');
-			$query->where('f.state=1');
+			$query->select($db->quoteName('DISTINCT *'));
+			$query->from($db->quoteName('#__tjfields_fields AS f'));
+			$query->where(
+				'NOT EXISTS (select * FROM '
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: use query builder to construct sub query

File: site/helpers/tjfields.php [hunk 10]
```diff
@@ -371,10 +359,10 @@ public function addFieldValueEntry($insert_obj)
 			$query = $db->getQuery(true);
 			$conditions = array($db->quoteName('id') . ' IN (' . $fieldValueEntryId . ') ');
 
-			$query->select("id")
-			->from("#__tjfields_options")
-			->where("field_id = " . $insert_obj->field_id)
-			->where("value = '" . $insert_obj->value . "'");
+			$query->select($db->quoteName('id'))
+			->from($db->quoteName('#__tjfields_options'))
+			->where($db->quoteName('field_id') . ' = ' . $db->quote($insert_obj->field_id))
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: no need to add $db->quote for integer value

**Comment by sagar-ttpl** [comment]
Work in this PR is in progress

**Pull request closed without merging**